src/twisted/internet/test/reactormixins.py

```diff
@@ -324,26 +326,30 @@ def stop():
             timedOutCall.cancel()
 
 
-    def makeTestCaseClasses(cls):
+    @classmethod
+    def makeTestCaseClasses(
+        cls: Type['ReactorBuilder']
+    ) -> Dict[str, Union[Type['ReactorBuilder'], Type['SynchronousTestCase']]]:
         """
         Create a L{SynchronousTestCase} subclass which mixes in C{cls} for each
         known reactor and return a dict mapping their names to them.
         """
-        classes = {}
+        classes = {}  # type: Dict[str, Union[Type['ReactorBuilder'], Type['SynchronousTestCase']]]   # noqa
         for reactor in cls._reactors:
             shortReactorName = reactor.split(".")[-1]
-            name = (cls.__name__ + "." + shortReactorName + "Tests").replace(".", "_")
-            class testcase(cls, SynchronousTestCase):
+            name = (cls.__name__ + "." + shortReactorName +
+                    "Tests").replace(".", "_")
+
+            class testcase(cls, SynchronousTestCase):  # type: ignore
```

Comment:
What doesn't mypy like about this?  Again, I would like type: ignores to be used sparingly and we should generally have a plan for how to get rid of them, at least file an upstream bug against mypy.

Reply:
If I didn't put that `#type: ignore` there, mypy would give:

```
src/twisted/internet/test/reactormixins.py:343:28: error: Variable "cls" is not valid as a type  [valid-type]
                class testcase(cls, SynchronousTestCase):
                               ^
src/twisted/internet/test/reactormixins.py:343:28: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
src/twisted/internet/test/reactormixins.py:343:28: error: Invalid base class "cls"  [misc]
                class testcase(cls, SynchronousTestCase):
```

The reason for this is that `cls` is a parameter, and the type of the parameter is only known at runtime.  So mypy doesn't know what type it is, and can't check it.

Reply:
Note that if you have to use `type: ignore`, you can specify the error category to ignore, which reduces the chance of accidentally silencing unexpected errors. In this case it would be `# type: ignore[valid-type]`.

Reply:
@mthuurne thanks for the feedback.  I've changed this line to specifically turn off two mypy errors on this line:

```
class testcase(cls, SynchronousTestCase):  # type: ignore[valid-type,misc]
```